### PR TITLE
Make sure streams always get closed

### PR DESCRIPTION
**Pull request by oxtoacart**
I'm going through the code trying to find conditions that could cause the stream leak seen in getlantern/lantern-internal#3336.

### CODE REVIEW

**Comment by oxtoacart** [comment]
There may be other conditions left, but this fix seems worth getting in for now.

File: listener.go [hunk 1]
```diff
@@ -202,6 +202,10 @@ func (l *listener) doOnConn(conn net.Conn) error {
 	})
 
 	consumeInboundTillDeadlineThenFail := func(fullErr error) error {
+		start := time.Now()
+		defer func() {
+			log.Debugf("Finished consuming inbound till deadline in %v", time.Now().Sub(start))
```
Comment (joesis) [inline listener.go]: Just FYI `time.Since(start)` can do this as well.

File: session.go [hunk 2]
```diff
@@ -522,6 +531,8 @@ func (s *session) onSessionError(readErr error, writeErr error) {
 	for _, c := range s.streams {
 		streams = append(streams, c)
 	}
+	// Clear out map of streams
+	s.streams = make(map[uint16]*stream, 0)
 	s.mx.RUnlock()
```
Comment (joesis) [inline session.go]: This should become `s.mx.Lock()` now

File: session.go [hunk 3]
```diff
@@ -633,6 +644,20 @@ func (s *session) Close() error {
 		atomic.AddInt64(&closingSessions, -1)
 		atomic.AddInt64(&openSessions, -1)
 		atomic.AddInt64(&closedSessions, 1)
+		go func() {
+			// wait until we're finished sending and receiving and then close any remaining streams
+			<-s.finishedSendingCh
+			<-s.finishedReceivingCh
+
+			s.mx.RLock()
```
Comment (joesis) [inline session.go]: This should be `s.mx.Lock()`

File: session.go [hunk 4]
```diff
@@ -633,6 +644,20 @@ func (s *session) Close() error {
...
+			for _, c := range s.streams {
+				// Note - we never send an RST because the underlying connection is
+				// considered no good at this point and we won't bother sending anything.
+				go c.close(false, nil, nil)
+			}
+			s.streams = nil
+			s.mx.RUnlock()
+		}()
```
Comment (joesis) [inline session.go]: I got unexpected test errors when trying to remove the duplicated code in `onSessionError`. In theory code here should supersede it by making sure the send and receive loop exits before closing sessions.

Comment (joesis) [reply session.go]: Okay have a PR https://github.com/getlantern/lampshade/pull/48 against this. Kindly let me know how you feel about that one.

**Comment by oxtoacart** [comment]
@joesis Thanks for the review!

**Pull request merged**